src/test/java/com/google/edith/LoginServletTest.java

```diff
@@ -0,0 +1,105 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.edith;
+
+import com.google.edith.servlets.LoginServlet;
+
+import com.google.appengine.api.users.User;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.tools.development.testing.LocalUserServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mock;
+import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.atLeast;
+import org.mockito.MockitoAnnotations;
+import org.mockito.ArgumentCaptor;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.io.IOException;
+import java.io.StringWriter;
+import java.io.PrintWriter;
+
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+public final class LoginServletTest {
+  
+  private Map<String, Object> myMap = new HashMap<String, Object>() {{
+        put("com.google.appengine.api.users.UserService.user_id_key", "12345");
+    }};
+
+  private final LocalServiceTestHelper loggedInTestHelper =
+      new LocalServiceTestHelper(new LocalUserServiceTestConfig())
+      .setEnvAttributes(myMap)
+      .setEnvIsLoggedIn(true)
+      .setEnvAuthDomain("gmail")
+      .setEnvIsAdmin(true)
+      .setEnvEmail("user@gmail.com");
+      
+  private final LocalServiceTestHelper loggedOutTestHelper =
+      new LocalServiceTestHelper(new LocalUserServiceTestConfig())
+      .setEnvIsLoggedIn(false);
+  
+  private final UserService userService = UserServiceFactory.getUserService();
+  private LoginServlet loginServlet;
+
+  @Mock
+  HttpServletRequest request;
+
+  @Mock
+  HttpServletResponse response;
+
+  @Before
+  public void setUp() throws Exception {
+    MockitoAnnotations.initMocks(this);
+    loginServlet = new LoginServlet();
+  }
+
+  @Test
+  public void testWhenLoggedIn() throws IOException, ServletException {
```

Comment:
Ideally test names should include the condition being testing and the expected outcome. E.g.
"multiply_twoNegativeValues_returnsPositiveValues". This makes isolating failure while debugging and pushing new changes a lot easier and faster.

Could we update the test names in this class?

Reply:
Most Definitely.
Changes made.